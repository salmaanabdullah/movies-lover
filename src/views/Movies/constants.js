import { asyncActionType } from 'src/utils/reduxActions';

export const NAMESPACE = 'MOVIES';

const FETCH_MOVIES = asyncActionType(`${NAMESPACE} FETCH_MOVIES`);
const FETCH_MOVIE_DETAILS = asyncActionType(`${NAMESPACE} FETCH_MOVIE_DETAILS`);
const RESET_DETAILS = `${NAMESPACE} RESET_DETAILS`;

export const actionsTypes = {
  FETCH_MOVIES,
  RESET_DETAILS,
  FETCH_MOVIE_DETAILS,
};
