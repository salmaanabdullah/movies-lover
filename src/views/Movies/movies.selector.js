import createSelector from 'src/utils/reselect';
import { List, Map } from 'immutable';

export const getMovies = state => state.get('movies', new Map());

export const getMoviesStatus = createSelector(getMovies, data => data.get('status'));
export const getMoviesList = createSelector(getMovies, data => data.get('list') || new List());
export const getMovieDetails = createSelector(getMovies, data => data.get('details') || new Map());
