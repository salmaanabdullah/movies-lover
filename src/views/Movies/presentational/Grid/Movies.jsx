import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import ImmutablePropTypes from 'react-immutable-proptypes';
import { Row, Col, Container } from 'react-bootstrap';
import MovieCard from '../MovieCard';
import styles from './style.module.scss';

const Movies = props => {
  const { getMovies, movies } = props;

  useEffect(() => {
    getMovies();
  }, [getMovies]);

  // useEffect(() => {
  //   throw new Error('Oh no!'); // todo if you want to test ErrorBoundry
  // }, []);

  const movieColumns = movies
    ? movies.map(movie => (
        <Col className={styles.movieColumn} key={movie.get('id')} xs={12} sm={6} md={4} lg={3}>
          <MovieCard movie={movie} />
        </Col>
      ))
    : null;

  return (
    <Container fluid>
      <Row>{movieColumns}</Row>
    </Container>
  );
};

Movies.propTypes = {
  getMovies: PropTypes.func.isRequired,
  movies: ImmutablePropTypes.list.isRequired,
};

export default Movies;
