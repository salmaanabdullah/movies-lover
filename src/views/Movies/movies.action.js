import { actionsTypes } from './constants';

export const getMovies = {
  pending: () => ({
    type: actionsTypes.FETCH_MOVIES.PENDING,
  }),
  success: ({ page, results, total_pages, total_results }) => ({
    type: actionsTypes.FETCH_MOVIES.SUCCESS,
    payload: { page, results, total_pages, total_results },
  }),
  error: () => ({
    type: actionsTypes.FETCH_MOVIES.ERROR,
  }),
};

export const getMovieDetailsAction = {
  pending: id => ({
    type: actionsTypes.FETCH_MOVIE_DETAILS.PENDING,
    payload: { id },
  }),
  success: details => ({
    type: actionsTypes.FETCH_MOVIE_DETAILS.SUCCESS,
    payload: { details },
  }),
  error: () => ({
    type: actionsTypes.FETCH_MOVIE_DETAILS.ERROR,
  }),
};

export const resetDetails = () => ({
  type: actionsTypes.RESET_DETAILS,
});
