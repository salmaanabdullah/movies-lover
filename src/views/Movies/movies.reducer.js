import { fromJS, Map } from 'immutable';
import { STATUS } from 'src/constants/Status';
import { actionsTypes } from './constants';

const initialState = fromJS({
  status: null,
  page: 0,
  list: [],
  total_pages: 0,
  total_results: 0,
  detailsStatus: null,
  details: {},
});

const moviesReducer = (state = initialState, action) => {
  const { type, payload } = action;
  switch (type) {
    case actionsTypes.FETCH_MOVIES.PENDING:
      return state.set('status', STATUS.LOADING);
    case actionsTypes.FETCH_MOVIES.SUCCESS:
      return state
        .set('status', STATUS.SUCCESS)
        .set('page', payload.page)
        .set('list', fromJS(payload.results))
        .set('total_pages', payload.total_pages)
        .set('total_results', payload.total_results);
    case actionsTypes.FETCH_MOVIES.ERROR:
      return state.set('status', STATUS.ERROR);

    case actionsTypes.FETCH_MOVIE_DETAILS.PENDING:
      return state.set('detailsStatus', STATUS.LOADING);
    case actionsTypes.FETCH_MOVIE_DETAILS.SUCCESS:
      return state.set('detailsStatus', STATUS.SUCCESS).set('details', new Map(payload.details));
    case actionsTypes.FETCH_MOVIE_DETAILS.ERROR:
      return state.set('detailsStatus', STATUS.ERROR);

    case actionsTypes.RESET_DETAILS:
      return state.set('details', new Map());

    default:
      return state;
  }
};
export default moviesReducer;
