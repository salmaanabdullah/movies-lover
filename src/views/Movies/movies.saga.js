import { all, takeLatest, put, call } from 'redux-saga/effects';
import { Movies } from 'src/services/movies';
import { actionsTypes } from './constants';
import { getMovieDetailsAction, getMovies } from './movies.action';

export function* getMoviesData() {
  try {
    const response = yield call(Movies.get);
    const { page, results, total_pages, total_results } = response.data;
    yield put(
      getMovies.success({
        page,
        results,
        total_pages,
        total_results,
      }),
    );
  } catch (error) {
    console.error(error); // eslint-disable-line
    yield put(getMovies.error());
  }
}
export function* getMoviesDetails({ payload }) {
  const { id } = payload;
  try {
    const response = yield call(Movies.getById, id);
    yield put(getMovieDetailsAction.success(response.data));
  } catch (error) {
    console.error(error); // eslint-disable-line
    yield put(getMovieDetailsAction.error());
  }
}

export default function* appSettingsSagas() {
  yield all([
    takeLatest(actionsTypes.FETCH_MOVIES.PENDING, getMoviesData),
    takeLatest(actionsTypes.FETCH_MOVIE_DETAILS.PENDING, getMoviesDetails),
  ]);
}
