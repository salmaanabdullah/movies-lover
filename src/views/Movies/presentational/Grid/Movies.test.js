import React from 'react';
import { List } from 'immutable';
import { shallow } from 'enzyme';
import Movies from './Movies';

it('expect to render movies component', () => {
  expect(shallow(<Movies movies={new List()} getMovies={jest.fn()}></Movies>)).toMatchSnapshot();
});
