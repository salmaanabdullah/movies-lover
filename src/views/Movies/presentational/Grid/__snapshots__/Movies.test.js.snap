// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect to render movies component 1`] = `
<Container
  fluid={true}
>
  <Row />
</Container>
`;
